mutiny-core/src/nostr/nwc.rs

```diff
@@ -155,94 +175,111 @@ impl NostrWalletConnect {
-                    .unwrap_or_default();
-
-                current.push(pending);
-
-                node_manager
-                    .storage
-                    .set_data(PENDING_NWC_EVENTS_KEY, current)?;
-
-                return Ok(None);
-            } else {
-                if invoice.amount_milli_satoshis().is_none() {
-                    log_error!(
-                        node_manager.logger,
-                        "NWC Invoice amount not set, cannot pay: {invoice}"
-                    );
-                    return Ok(None);
-                }
-
-                let msats = invoice.amount_milli_satoshis().unwrap();
-
-                // verify amount is under our limit
-                let content = if msats <= self.profile.max_single_amt_sats * 1_000 {
-                    self.pay_nwc_invoice(node_manager, from_node, &invoice)
-                        .await
-                } else {
-                    log_warn!(
-                        node_manager.logger,
-                        "Invoice amount too high: {msats} msats"
-                    );
-
-                    Response {
-                        result_type: Method::PayInvoice,
-                        error: Some(NIP47Error {
-                            code: ErrorCode::QuotaExceeded,
-                            message: format!("Invoice amount too high: {msats} msats"),
-                        }),
-                        result: None,
+            match self.profile.spending_conditions.clone() {
+                SpendingConditions::SingleUse(mut single_use) => {
+                    // check if we have already spent
+                    if single_use.spent {
+                        return Ok((None, needs_save));
                     }
-                };
-
-                let encrypted = encrypt(&server_key, &client_pubkey, content.as_json())?;
 
-                let p_tag = Tag::PubKey(event.pubkey, None);
-                let e_tag = Tag::Event(event.id, None, None);
-                let response =
-                    EventBuilder::new(Kind::WalletConnectResponse, encrypted, &[p_tag, e_tag])
-                        .to_event(&self.server_key)?;
+                    // todo abstract paying invoice to a separate function
+                    if invoice.amount_milli_satoshis().is_none() {
+                        log_error!(
+                            node_manager.logger,
+                            "NWC Invoice amount not set, cannot pay: {invoice}"
+                        );
+                        return Ok((None, needs_save));
+                    }
 
-                return Ok(Some(response));
+                    let msats = invoice.amount_milli_satoshis().unwrap();
+
+                    // verify amount is under our limit
+                    let content = if msats <= single_use.amount_sats * 1_000 {
+                        let response = self
+                            .pay_nwc_invoice(node_manager, from_node, &invoice)
```

Comment:
we could have a race here where we get multiple requests at the same time and pay all of them before we update that this nwc is spent.

Reply:
this is actually fine because we don't process in parallel